pylint/message/message.py

```diff
@@ -4,90 +4,78 @@
+from dataclasses import asdict, dataclass
 from warnings import warn
 
 from pylint.constants import MSG_TYPES
-from pylint.interfaces import Confidence
+from pylint.interfaces import UNDEFINED, Confidence
 from pylint.typing import MessageLocationTuple
 
-_MsgBase = collections.namedtuple(
-    "_MsgBase",
-    [
-        "msg_id",
-        "symbol",
-        "msg",
-        "C",
-        "category",
-        "confidence",
-        "abspath",
-        "path",
-        "module",
-        "obj",
-        "line",
-        "column",
-        "end_line",
-        "end_column",
-    ],
-)
 
-
-class Message(_MsgBase):
+@dataclass(unsafe_hash=True)
+class Message:  # pylint: disable=too-many-instance-attributes
     """This class represent a message to be issued by the reporters."""
 
-    @overload
-    def __new__(
-        cls,
-        msg_id: str,
-        symbol: str,
-        location: MessageLocationTuple,
-        msg: str,
-        confidence: Confidence | None,
-    ) -> Message:
-        ...
-
-    @overload
-    def __new__(
-        cls,
-        msg_id: str,
-        symbol: str,
-        location: tuple[str, str, str, str, int, int],
-        msg: str,
-        confidence: Confidence | None,
-    ) -> Message:
-        # Remove for pylint 3.0
-        ...
+    msg_id: str
+    symbol: str
+    msg: str
+    C: str
+    category: str
+    confidence: Confidence
+    abspath: str
+    path: str
+    module: str
+    obj: str
+    line: int
+    column: int
+    end_line: int | None
+    end_column: int | None
+    location: MessageLocationTuple
 
-    def __new__(
-        cls,
+    def __init__(
+        self,
         msg_id: str,
         symbol: str,
         location: tuple[str, str, str, str, int, int] | MessageLocationTuple,
         msg: str,
         confidence: Confidence | None,
-    ) -> Message:
+    ) -> None:
         if not isinstance(location, MessageLocationTuple):
             warn(
                 "In pylint 3.0, Messages will only accept a MessageLocationTuple as location parameter",
                 DeprecationWarning,
             )
-            location = location + (None, None)  # type: ignore[assignment] # Temporary fix until deprecation
-        return _MsgBase.__new__(
-            cls,
-            msg_id,
-            symbol,
-            msg,
-            msg_id[0],
-            MSG_TYPES[msg_id[0]],
-            confidence,
-            *location,
-        )
+            location = MessageLocationTuple(
+                location[0],
+                location[1],
+                location[2],
+                location[3],
+                location[4],
+                location[5],
+                None,
+                None,
+            )
+
+        self.msg_id = msg_id
+        self.symbol = symbol
+        self.msg = msg
+        self.C = msg_id[0]
+        self.category = MSG_TYPES[msg_id[0]]
+        self.confidence = confidence or UNDEFINED
+        self.abspath = location.abspath
```

Comment:
We could use the location as an attribute and create property to not break the API.

Reply:
Not sure what you mean, see L73. I added location to be able to access it directly and with the new `if` we can now assume it is `MessageLocationTuple` here.

Reply:
Doesn't it mean that we could modify "line" without modifying "location.line" ? Maybe we need to use a frozen dataclass to be sure it won't be incoherent.

Reply:
Ah, yeah I didn't consider that!

I'll remove the `location` attribute. We don't even need it ourselves. That makes it easier. If we want to expand the API later we can think about it then.